File: 000601-000800/000724.寻找数组的中心下标/724.js
/**
 * @param {number[]} nums
 * @return {number}
 */
// 方法一：用时 88 ms
// 执行用时超过了 61 %
// 消耗内存超过了 85 %
// 难度: 简单
// 同: https://leetcode-cn.com/problems/find-the-middle-index-in-array
// 同: https://github.com/vjudge/leetcode/tree/master/001801-002000/001991.找到数组的中间位置
var pivotIndex = function(nums) {
    let sum = 0
    for (const num of nums) {
        sum += num
    }
    let midsum = 0
    for (let i = 0; i < nums.length; i++) {
        if (2 * midsum + nums[i] == sum) {
            return i
        }
        midsum += nums[i]
    }
    return -1
}

console.log(pivotIndex([1, 7, 3, 6, 5, 6]))
console.log(pivotIndex([1, 2, 3]))
console.log(pivotIndex([2, 1, -1]))
console.log(pivotIndex([2, 3, -1, 8, 4]))
console.log(pivotIndex([1, -1, 4]))
console.log(pivotIndex([2, 5]))
console.log(pivotIndex([1]))

File: 000601-000800/000724.寻找数组的中心下标/724.go
package main

import "fmt"

// 方法一：用时 20 ms
// 执行用时超过了 71 %
// 消耗内存超过了 67 %
// 难度: 简单
// 同: https://leetcode-cn.com/problems/find-the-middle-index-in-array
// 同: https://github.com/vjudge/leetcode/tree/master/001801-002000/001991.找到数组的中间位置
func pivotIndex(nums []int) int {
	sum := 0
	for _, num := range nums {
		sum += num
	}
	midsum := 0
	for i, num := range nums {
		if 2 * midsum + num == sum {
			return i
		}
		midsum += num
	}
	return -1
}

func main ()  {
	fmt.Println(pivotIndex([]int{1, 7, 3, 6, 5, 6}))
	fmt.Println(pivotIndex([]int{1, 2, 3}))
	fmt.Println(pivotIndex([]int{2, 1, -1}))
	fmt.Println(pivotIndex([]int{2, 3, -1, 8, 4}))
	fmt.Println(pivotIndex([]int{1, -1, 4}))
	fmt.Println(pivotIndex([]int{2, 5}))
	fmt.Println(pivotIndex([]int{1}))
}